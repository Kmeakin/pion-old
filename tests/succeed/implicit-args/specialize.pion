//~ run = "${PION} elab ${FILE}"

let id = fun (@A : Type) (a : A) => a;
let always = fun (@A : Type) (@B : Type) (a : A) (b : B) => a;
let apply = fun (@A : Type) (@B : Type) (f : A -> B) (x : A) => f x;

// No specialization
let _ = id;
let _ = always;
let _ = apply;

// Full specialization
let _ : Bool -> Bool = id;
let _ : Bool -> Int -> Bool = always;
let _ : (Bool -> Int) -> Bool -> Int = apply;

// Specialization of higher order functions
let _ = apply (always false) (0 : Int);

{}
