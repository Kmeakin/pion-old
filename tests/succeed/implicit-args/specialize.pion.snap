exit status: 0
stdout:
{
    let id : fun (@A : Type) -> A -> A = fun (@A : Type) (a : A) => a;
    let always : fun (@A : Type) (@B : Type) -> A -> B -> A = fun (@A : Type)
    (@B : Type)
    (a : A)
    (b : B) => a;
    let apply : fun (@A : Type) (@B : Type) -> (A -> B) -> A ->
    B = fun (@A : Type) (@B : Type) (f : A -> B) (x : A) => f x;
    let _ : fun (@A : Type) -> A -> A = id;
    let _ : fun (@A : Type) (@B : Type) -> A -> B -> A = always;
    let _ : fun (@A : Type) (@B : Type) -> (A -> B) -> A -> B = apply;
    let _ : Bool -> Bool = id @Bool;
    let _ : Bool -> Int -> Bool = always @Bool @Int;
    let _ : (Bool -> Int) -> Bool -> Int = apply @Bool @Int;
    let _ : Bool = apply @Int @Bool (always @Bool @Int false) 0;
    ()
} : ()

stderr:
